### PR TITLE
Restore check when using the SQL manager part

### LINKED ISSUE #24209: Error when using delete request in back office
#### Describe the bug

Error when using delete request in back office

#### Expected behavior
It shouldn't be possible to delete something using sql manager in back office

#### Steps to Reproduce

1. Go to admin -> database (with mode debug off)
2. Add a sql request using delete and save
3. Execute this request
4. See error and check in database that the request is executed.
--------------------------------------------------------------------------
1. Go to admin -> database (with mode debug on)
2. Add a sql request using delete and save
3. Execute this request
4. See error and check in database that the request isn't executed.

Before the migration it wasn't possible to save delete request.

**Screenshots**

Before migration (<1.7.6.0)  : https://drive.google.com/file/d/1QY8vPiWKlDbHPiwScdxyPk7ztL4Uoe-a/view
After migration (>1.7.6.0) : https://drive.google.com/file/d/17PTrAf2gqiM7dyQ_x1gJScsVrZXW44nZ/view

#### Additional information

* PrestaShop version: 1.7.6.9, 1.7.7.3, 1.7.8.x, develop
* PHP version: 7.1

### PR DESCRIPTION
**Pull request by PierreRambaud**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 1.7.7.x
| Description?      | The SQL manager is currently malformed and the validator is not executed. In fact, only the SELECT query should be allowed.
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | yes
| Fixed ticket?     | Fixes #24209 
| How to test?      | Only the SELECT can be executed.
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

## BC breaks:

* `handleDeleteException` is now called `handleException`, the name was wrong

## Deprecated 

* `PrestaShop\PrestaShop\Adapter\SqlManager\SqlRequestFormDataValidator`  use `prestashop.adapter.sql_manager.sql_query_validator` or `PrestaShop\PrestaShop\Adapter\SqlManager\SqlQueryValidator`
* `PrestaShopBundle\Form\Admin\Configure\AdvancedParameters\RequestSql\SqlRequestFormDataProvider` use `prestashop.core.form.identifiable_object.data_provider.sql_request_form_data_provider` or `PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataProvider\SqlRequestFormDataProvider`
* `PrestaShopBundle\Form\Admin\Configure\AdvancedParameters\RequestSql\SqlRequestFormHandler` use `PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataHandler\SqlRequestFormDataHandler`  or `prestashop.core.form.identifiable_object.sql_request_form_data_handler`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24217)

### CODE REVIEW

**Review by sowbiba — DISMISSED** [review]
LGTM. Cs-fixer not happy because of a blank line

File: src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php [hunk 1]
```diff
@@ -73,4 +95,22 @@ public function update($id, array $data)
...
+                $errors[0]['parameters'],
+                $errors[0]['domain']
+            );
+
+            throw new SqlRequestConstraintException(
+                $message,
+                SqlRequestConstraintException::INVALID_SQL_QUERY
+            );
+        }
+
```
Comment (sowbiba) [inline src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php]:
```suggestion
```

**Review by sowbiba — COMMENTED** [review]
blank line

**Review by sowbiba — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php [hunk 2]
```diff
@@ -173,19 +174,24 @@ public function processFormAction(Request $request)
...
+                $this->addFlash('success', $this->trans('Successful creation.', 'Admin.Notifications.Success'));
 
-            return $this->redirectToRoute('admin_sql_requests_index');
+                return $this->redirectToRoute('admin_sql_requests_index');
+            }
+        } catch (SqlRequestException $e) {
+            $this->addFlash('error', $this->handleDeleteException($e));
         }
 
+
```
Comment (sowbiba) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php]:
```suggestion
```

**Review by sowbiba — DISMISSED** [review]

File: src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php [hunk 3]
```diff
@@ -41,19 +44,36 @@ final class SqlRequestFormDataHandler implements FormDataHandlerInterface
...
+     * @var TranslatorInterface
+     */
+    private $translator;
+
     /**
      * @param CommandBusInterface $bus
      */
-    public function __construct(CommandBusInterface $bus)
-    {
+    public function __construct(
```
Comment (Progi1984) [inline src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php]: Could you add behat tests for this behavior change ?

File: src/Adapter/SqlManager/SqlQueryValidator.php [hunk 4]
```diff
@@ -84,10 +84,6 @@ private function getErrors(array $sqlErrors)
         $errors = [];
 
         foreach ($sqlErrors as $key => $sqlError) {
```
Comment (Progi1984) [inline src/Adapter/SqlManager/SqlQueryValidator.php:86]: Could you add unit tests for this behavior change ?

**Review by Progi1984 — COMMENTED** [review]
Small comments

File: src/Adapter/SqlManager/SqlQueryValidator.php (hunk 4 above)
Comment (PierreRambaud) [reply src/Adapter/SqlManager/SqlQueryValidator.php:86]: I can't, it's an ObjectModel validator :(

File: src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php (hunk 3 above)
Comment (PierreRambaud) [reply src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php]: I add simple behats tests for the form handling SELECT, UPDATE, DELETE, INSERT, Bad query, and unknown table.

File: tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        Assert::assertEquals(
+            '"SELECT" does not exist.',
+            $this->lastException->getMessage()
+        );
+    }
+
+    /**
+     * @Then I should get an error that the SQL request is malformed
+     */
+    public function assertLastErrorIsAMalformedSqlREquest()
```
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php]:
```suggestion
    public function assertLastErrorIsAMalformedSqlRequest(): void
```

File: src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php (hunk 3 above)
Comment (Progi1984) [reply src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php]: :+1:

**Review by Progi1984 — COMMENTED** [review]
Some pétouilles and it's good

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php [hunk 6]
```diff
@@ -173,17 +174,21 @@ public function processFormAction(Request $request)
...
+            $sqlRequestForm = $this->getSqlRequestFormBuilder()->getForm($data);
+            $sqlRequestForm->handleRequest($request);
 
-        $result = $this->getSqlRequestFormHandler()->handle($sqlRequestForm);
+            $result = $this->getSqlRequestFormHandler()->handle($sqlRequestForm);
 
-        if (null !== $result->getIdentifiableObjectId()) {
-            $this->addFlash('success', $this->trans('Successful creation.', 'Admin.Notifications.Success'));
+            if (null !== $result->getIdentifiableObjectId()) {
+                $this->addFlash('success', $this->trans('Successful creation.', 'Admin.Notifications.Success'));
```
Comment (Julievrz) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php:186]:
```suggestion
                $this->addFlash('success', $this->trans('Successful creation', 'Admin.Notifications.Success'));
```

Comment (PierreRambaud) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php:186]: I'm sorry, it's not related to the PR, I didn't change any wording :(

File: src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php [hunk 7]
```diff
@@ -41,19 +44,36 @@ final class SqlRequestFormDataHandler implements FormDataHandlerInterface
...
+        CommandBusInterface $bus,
+        SqlRequestFormDataValidator $requestSqlFormDataValidator,
+        TranslatorInterface $translator
+    ) {
         $this->commandBus = $bus;
+        $this->requestSqlFormDataValidator = $requestSqlFormDataValidator;
```
Comment (matks) [inline src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php]: If I am not wrong, this change in the constructor is a BC Break.

This class is `final` but if someone creates this class manually
```
$myService = new SqlRequestFormDataHandler($bus);
```
Then following this PR, his code will crash?

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php (hunk 6 above)
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php:186]: I confirm, @PierreRambaud only changed the line indentation 😉

File: src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormDataProvider.php [hunk 8]
```diff
@@ -38,8 +38,18 @@
 use PrestaShop\PrestaShop\Core\Domain\SqlManagement\Query\GetSqlRequestForEditing;
 use PrestaShop\PrestaShop\Core\Domain\SqlManagement\ValueObject\SqlRequestId;
 
+@trigger_error(
+    sprintf(
+        '%s is deprecated since version 1.7.7.4 and will be removed in the next major version.',
```
Comment (matks) [inline src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormDataProvider.php]: 1.7.7.5

File: src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormDataProvider.php [hunk 9]
```diff
@@ -38,8 +38,18 @@
...
+        SqlRequestFormDataProvider::class
+    ),
+    E_USER_DEPRECATED
+);
+
 /**
  * Class RequestSqlFormDataProvider is responsible for getting/saving RequestSql form data.
+ *
+ * @deprecated Since 1.7.7.4 and will be removed in the next major.
```
Comment (matks) [inline src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormDataProvider.php]: 1.7.7.5

File: src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormHandler.php [hunk 10]
```diff
@@ -30,8 +30,18 @@
 use Symfony\Component\Form\FormFactoryInterface;
 use Symfony\Component\Form\FormInterface;
 
+@trigger_error(
+    sprintf(
+        '%s is deprecated since version 1.7.7.4 and will be removed in the next major version.',
```
Comment (matks) [inline src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormHandler.php]: 1.7.7.5

File: src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormHandler.php [hunk 11]
```diff
@@ -30,8 +30,18 @@
...
+        SqlRequestFormHandler::class
+    ),
+    E_USER_DEPRECATED
+);
+
 /**
  * Class RequestSqlFormHandler is responsible for creating RequestSql form.
+ *
+ * @deprecated Since 1.7.7.4 and will be removed in the next major.
```
Comment (matks) [inline src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormHandler.php]: 1.7.7.5

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_handler.yml [hunk 12]
```diff
@@ -335,19 +335,21 @@ services:
...
 
-  # Entity form handler
+    # Entity form handler
     prestashop.admin.request_sql.form_handler:
         class: 'PrestaShopBundle\Form\Admin\Configure\AdvancedParameters\RequestSql\SqlRequestFormHandler'
         arguments:
             - '@form.factory'
             - '@prestashop.adapter.sql_manager.request_sql_form_data_validator'
             - '@prestashop.bundle.form.admin.configure.advanced_parameters.request_sql.form_data_provider'
+        deprecated: 'The "%service_id% service is deprecated. Use "prestashop.core.form.identifiable_object.sql_request_form_data_handler" instead'
```
Comment (matks) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_handler.yml:345]: 👏

File: tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    {
+        $data = $node->getRowsHash();
+
+        SharedStorage::getStorage()->set(static::SQL_REQUEST_FORM_KEY, $data);
+    }
+
+    /**
+     * @When I submit the sql request form
+     */
+    public function submitTheSqlRequestForm(): void
```
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php]: I do not understand what you do in this test.

You modified the behavior of the class SqlRequestFormDataHandler. The class SqlRequestFormDataHandler has a very simple API. Why not test the class SqlRequestFormDataHandler?

Why write a test that involves services `prestashop.core.form.builder.sql_request_form_builder` and `prestashop.core.form.identifiable_object.sql_request_form_handler` even though they are not concerned?

Morever, service `prestashop.core.form.builder.sql_request_form_builder` is a FormBuilder service that belongs to the application layer. In order to test it, you have to fake creating a request using ` $request = Request::createFromGlobals();`. I think it indicates this test does not belong to this layer.

Our Behat tests are supposed to test _behaviors_ of our code, our domain logic. Our domain logic should - as much as possible - be unaware whether behaviors are done through a web interface or a CLI.

I think if we write some Behat tests that depend on a service which has "form" in its name, we are doing something wrong. I am sorry but I think this is not the right test layer.

**Review by matks — DISMISSED** [review]

File: src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php (hunk 7 above)
Comment (PierreRambaud) [reply src/Core/Form/IdentifiableObject/DataHandler/SqlRequestFormDataHandler.php]: The fact is, we don't consider final class as BC for months :sweat: 
A lot of BC breaks have forgotten because of this :(

File: src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormDataProvider.php (hunk 8 above)
Comment (PierreRambaud) [reply src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/RequestSql/SqlRequestFormDataProvider.php]: I was thinkg it will be merged for the 1.7.7.4 :(

File: tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php (hunk 13 above)
Comment (PierreRambaud) [reply tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php]: > You modified the behavior of the class SqlRequestFormDataHandler. The class SqlRequestFormDataHandler has a very simple API. Why not test the class SqlRequestFormDataHandler?

I want to test the whole process and validation.

> Why write a test that involves services prestashop.core.form.builder.sql_request_form_builder and prestashop.core.form.identifiable_object.sql_request_form_handler even though they are not concerned?

It's an integration test, I am testing the whole process, as it is done in the Core.

> Morever, service prestashop.core.form.builder.sql_request_form_builder is a FormBuilder service that belongs to the application layer. In order to test it, you have to fake creating a request using $request = Request::createFromGlobals();. I think it indicates this test does not belong to this layer.

I see no problem creating fake request for purpose :thinking: 

> Our Behat tests are supposed to test behaviors of our code, our domain logic. Our domain logic should - as much as possible - be unaware whether behaviors are done through a web interface or a CLI.

Behat is a tool that makes behavior driven development (BDD) possible with human-readable stories that describe the behavior of the application. This is exactly what I'm doing here :thinking:

**Review by Progi1984 — DISMISSED** [review]

File: tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php (hunk 13 above)
Comment (matks) [reply tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php]: Let's cut long discussions and ask @eternoendless his opinion about this test

**Review by sowbiba — DISMISSED** [review]

File: tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php (hunk 13 above)
Comment (sowbiba) [reply tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php]: According to what @matks said, If we want make it testable as we used to do it, the SQL request validation have to be performed in the AddSqlRequestHandler and EditSqlRequestHandler
In this way we will be able to test without the formHandler and the whole request

Comment (eternoendless) [reply tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php]: We have several kinds of tests, and the frontier between them is not self-evident, so it's understandable that there might be confusion and a diversity of opinions.

To me it all comes down to what are you trying to test:

- Are you testing what the user sees? While Behat could be an adequate tool for that, we use UI tests to test that.
- Are you testing the domain behavior? Go for Behat.
- Are you testing several components together? Integration (phpunit)
- Are you testing a single component? Unit

Remember the domain enforces validation so that people cannot do with it what they're not supposed to. Forms on the other hand are meant to provide a user interface to the domain, and while they do perform data validation, it is there to provide feedback about what exactly the user did wrong, not as a barrier to protect the system (that is the domain's job).

In conclusion, as @sowbiba said, I think your behavior tests should be checking domain exceptions thrown by `AddSqlRequestHandler` and `EditSqlRequestHandler`, not the form handler. If you want to test the actual error messages sent by the form handler, it would be more appropriate to test that with a UI test.

Both approaches will lead to test cases that are simpler, more robust, and easier to maintain.

**Comment by PierreRambaud** [comment]
Regarding the latest conversation with @eternoendless, I have to edit and tests `AddSqlRequestHandler` and `EditSqlRequestHandler` which are currently saving everything without any check (compared to the form).

File: src/Adapter/SqlManager/CommandHandler/AbstractSqlRequestHandler.php [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\SqlManager\CommandHandler;
```
Comment (Progi1984) [inline src/Adapter/SqlManager/CommandHandler/AbstractSqlRequestHandler.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Adapter\SqlManager\CommandHandler;
```

File: src/Adapter/SqlManager/CommandHandler/AbstractSqlRequestHandler.php [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+use PrestaShop\PrestaShop\Adapter\SqlManager\SqlQueryValidator;
+use PrestaShop\PrestaShop\Core\CommandBus\CommandBusInterface;
+use PrestaShop\PrestaShop\Core\Domain\SqlManagement\Exception\SqlRequestConstraintException;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * Class EditSqlRequestHandler is responsible for updating SqlRequest.
+ *
```
Comment (Progi1984) [inline src/Adapter/SqlManager/CommandHandler/AbstractSqlRequestHandler.php]:
```suggestion
```

**Review by Progi1984 — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php [hunk 16]
```diff
@@ -471,13 +478,14 @@ protected function getSqlRequestDataFromRequest(Request $request)
      *
      * @return string Error message
      */
-    protected function handleDeleteException(SqlRequestException $e)
+    protected function handleException(SqlRequestException $e)
```
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php:481]: Could you explain the BC Break

**Review by Progi1984 — COMMENTED** [review]
A small request

**Review by Progi1984 — COMMENTED** [review]
A last request for documentation and it's good

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php (hunk 6 above)
Comment (Julievrz) [reply src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/SqlManagerController.php:186]: Hi, I'm removing the 'waiting for wording' label then

**Review by atomiix — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]

**Comment by sarahdib** [comment]
@PierreRambaud 
It's impossible with your pr to add a delete request : ✅ 
But if the delete request was here before the issue #24209 is not fixed. Is it possible to fix this case ?

**Comment by PierreRambaud** [comment]
> @PierreRambaud
> It's impossible with your pr to add a delete request : 
> But if the delete request was here before the issue #24209 is not fixed. Is it possible to fix this case ?

Unfortunately no, because you shouldn't have been able to create delete request. This PR aims to restore the normal behavior, meaning only SELECT request are authorized.
